lib/spack/spack/build_environment.py

```diff
@@ -775,6 +775,15 @@ def parent_class_modules(cls):
     return result
 
 
+def nonshared_parents(cls):
+    """Return list with cls and all superclasses that define a Package
+       class object with the same name (so excludes superclasses like
+       PackageBase).
+    """
+    all_parents = parent_class_modules(cls)
+    return list(x for x in all_parents if getattr(x, cls.__name__, None))
```

Comment:
I don't think we want to exclude name changes -- what if someone has `my-hdf5` that shares everything except a URL with the builtin hdf5 -- they should be able to define a package `class MyHdf5(Hdf5):`.

What if instead we chose only classes that come from files named `package.py`? that should reasonably-effectively get only package classes, not base classes?

Reply:
That would work IMO but would run into the same problem as my other suggestion:

> this would cause problems if users created packages in their repo for the sole purpose of sharing amongst other packages